### DISCUSSION TITLE
Validate properties of records with the init keyword when using the with expressions

### DISCUSSION THREAD

**Discussion by ChristophvdF** [dotnet/csharplang#5652] [Language Ideas]
Currently I can validate a records as followed

``` csharp
public record PersonWithValidation(string FirstName, string LastName)
{
    public string FirstName { get; init; } = FirstName ?? throw new ArgumentNullException(nameof(FirstName));

    public string LastName { get; init; } = LastName ?? throw new ArgumentNullException(nameof(LastName));
}
```
When an instance is created with the `new` keyword validation will be executed. On the other hand, if I clone an instance using the `with` keyword the validation will be skipped. 

``` csharp
var original = new PersonWithValidation(firstName, lastName);
PersonWithValidation copy = original with { FirstName = null! }; // <-- First name is null
```

What was a bit unexpected.

The IL code of the above mentioned record reveled that the validation is added to the constructor like
``` csharp
string str1 = FirstName;
if (str1 == null)
     throw new ArgumentNullException(nameof (FirstName));
 // ISSUE: reference to a compiler-generated field
this.<FirstName>k__BackingField = str1;
```

If this validation will be added to the setter of the created property the validation will also be run when the object is cloned using the `with` keyword.

Simplified version to demonstrate my intent
``` csharp
public string FirstName 
{ 
    get
    {
        return this.<FirstName>k__BackingField
    } 
    set
    {
        if(value == null)
        {
            throw new ArgumentNullException(nameof (FirstName));
        }

        this.<FirstName>k__BackingField = value
    } 
}
```

The notation using `this.<FirstName>k__BackingField` is also used in the copy constructor of the record.
``` csharp
protected PersonWithValidation(PersonWithValidation original)
    {
      // ISSUE: reference to a compiler-generated field
      // ISSUE: reference to a compiler-generated field
      this.\u003CFirstName\u003Ek__BackingField = original.\u003CFirstName\u003Ek__BackingField;
      // ISSUE: reference to a compiler-generated field
      // ISSUE: reference to a compiler-generated field
      this.\u003CLastName\u003Ek__BackingField = original.\u003CLastName\u003Ek__BackingField;
    }
```

So I think this could be feasible.

**Comment by bernd5** [comment]
Currently you have to write code like this:

```csharp
using System;
#nullable enable

var original = new PersonWithValidation("Foo", "Bar");
PersonWithValidation copy = original with { FirstName = null! }; 

public record PersonWithValidation(string FirstName, string LastName)
{
    private string firstName = ValidateProperty(FirstName);
    public string FirstName 
    { 
        get => firstName;
        init{
            firstName = ValidateProperty(value, nameof(FirstName));
        }
    }

    private string lastName = ValidateProperty(LastName);
    
    public string LastName 
    { 
        get => lastName;
        init{
            lastName = ValidateProperty(value, nameof(LastName));
        }
    }
    
    private static T ValidateProperty<T>(T t,
        [System.Runtime.CompilerServices.CallerArgumentExpression("t")]string? name = null)
    {
        if (t == null){
             throw new ArgumentNullException(name);
        }
        return t;
    }
}
```
But it might become simpler with: [semi-auto-properties](https://github.com/dotnet/csharplang/issues/140)

**Reply by ChristophvdF** [reply]
There are two thing bugging my in the current solution.
1. I need a backing field
2. the `init` keyword represents a readonly property which can be 'skipped' by object initialization but the validation is skipped

My thought during development where. 
Ok, I want this to be readonly but I need validation while constructing the object. 

Utilizing the new features I truly thought that when using the `with` keyword the validation will still be used since the object is created in the same manner, so I thought. You are not able, or at least I don't know how, to break a readonly property like this and that was my thought on the `init` keyword. 
Since I found out that the validation is moved to the constructor for me, the next logical thing would be to move it to the setter of the property so it is always validated. Surely after digging into the matter and thinking of all the utilized features it was clear that I was thinking in the wrong direction, but I still believe that I would not be the only one making such a conclusion about validation of `init` properties.
I for sure will keep an eye on the semi-auto-properties, thanks for the link!

**Reply by HaloFour** [reply]
> 2. the `init` keyword represents a readonly property which can be 'skipped' by object initialization but the validation is skipped

[Required properties](https://github.com/dotnet/csharplang/issues/3630), which are also currently in design, would allow you to specify that the property must be initialized during creation.

**Reply by Joe4evr** [reply]
> Ok, I want this to be readonly but I need validation while constructing the object.

Of note: something that's also on the languge team's radar (though in unknown state at this time) [is some special "validator" method](https://youtu.be/fIXujTOA6BE?t=2749) which would be something that the compiler is compelled to call after object construction/initialization. Rather than calling your validation logic from each individual setter, you do it for all properties in sequence (and all in one place).
